nengo/tests/test_learning_rules.py

```diff
@@ -466,3 +467,118 @@ def test_frozen():
+            def regularized_pes(target, data):
+                return -dt / n * np.outer(data[:2], data[2:]) - target * reg
+        else:
+            def regularized_pes(target, data, params):
+                encoders = params[b].encoders
+                gains = params[b].gain
+                return (-dt / np.sqrt(n) *
+                        np.outer(np.dot(encoders, data[:2]) * gains,
+                                 data[2:]) - target * reg)
+
+        rule = GenericRule(regularized_pes, learning_rate=1e-3,
+                           size_in=2 + n, modifies=modifies,
+                           pass_model_params=modifies == "weights")
+
+        conn = nengo.Connection(
+            a if modifies == "decoders" else a.neurons,
+            b if modifies == "decoders" else b.neurons,
+            transform=1 if modifies == "decoders" else np.eye(n),
+            learning_rule_type=rule)
+        nengo.Connection(b, conn.learning_rule[:2], transform=1)
+        nengo.Connection(u, conn.learning_rule[:2], transform=1)
+        nengo.Connection(a.neurons, conn.learning_rule[2:], synapse=0.005)
+
+        b_p = nengo.Probe(b, synapse=0.1)
+        weights_p = nengo.Probe(conn, attr="weights")
+
+    sim = Simulator(m, dt=dt)
+    sim.run(1.)
+
+    tmask = sim.trange() > .9
+    assert (np.sum(np.abs(sim.data[weights_p][:10])) >
+            np.sum(np.abs(sim.data[weights_p][-10:])))
+    assert np.allclose(sim.data[b_p][tmask], -input_vector, atol=0.05)
+
+
+def test_generic_encoders(Simulator, nl_nodirect, rng, seed):
+    """Voja test implemented via generic rule."""
+
+    n = 200
+    learned_vector = np.asarray([0.3, -0.4, 0.6])
+    learned_vector /= np.linalg.norm(learned_vector)
+    n_change = n // 2  # modify first half of the encoders
+    dt = 1e-3
+
+    # Set the first half to always fire with random encoders, and the
+    # remainder to never fire due to their encoder's dot product with the input
+    intercepts = np.asarray([-1] * n_change + [0.99] * (n - n_change))
+    rand_encoders = UniformHypersphere(surface=True).sample(
+        n_change, len(learned_vector), rng=rng)
+    encoders = np.append(
+        rand_encoders, [-learned_vector] * (n - n_change), axis=0)
+
+    m = nengo.Network(seed=seed)
+    with m:
+        m.config[nengo.Ensemble].neuron_type = nl_nodirect()
+        u = nengo.Node(output=learned_vector)
+        x = nengo.Ensemble(n, dimensions=len(learned_vector),
+                           intercepts=intercepts, encoders=encoders,
+                           radius=2.0)  # to test encoder scaling
+
+        def generic_voja(enc, data, params):
+            scale = (params[x].gain / x.radius)[:, None]
+            return (scale * np.outer(data[:n], data[n:]) -
+                    data[:n, None] * enc)
+
+        conn = nengo.Connection(
+            u, x, synapse=None,
+            learning_rule_type=GenericRule(
+                generic_voja, learning_rate=1e-1 * dt, modifies="encoders",
```

Comment:
Indeed. I think it would be good to unify because the tests are also somewhat an example of what we recommend. So if we recommend to include the `dt` in the learning rule we should do that in both cases or if we don't that we shouldn't. (To me it seems the former case would be better as the `dt` is defined by the simulator and can change.) This should probably also agree with the specialized implementations of these learning rules.

Maybe the `dt` should be passed as argument to the generic learning rule function from the simulator (to avoid defining it independent of the simulator)?